curvefs/src/mds/topology/topology.cpp

```diff
@@ -0,0 +1,931 @@
+                if (!storage_->StoragePartition(data)) {
+                    return TopoStatusCode::TOPO_STORGE_FAIL;
+                }
+                partitionMap_[id] = data;
+                it->second.AddPartitionNum();
+                it->second.SetDirtyFlag(true);
+                return TopoStatusCode::TOPO_OK;
+            } else {
+                return TopoStatusCode::TOPO_ID_DUPLICATED;
+            }
+        } else {
+            return TopoStatusCode::TOPO_COPYSET_NOT_FOUND;
+        }
+    } else {
+        return TopoStatusCode::TOPO_POOL_NOT_FOUND;
+    }
+}
+
+TopoStatusCode TopologyImpl::RemovePartition(PartitionIdType id) {
+    WriteLockGuard wlockCopySet(copySetMutex_);
+    WriteLockGuard wlockPartition(partitionMutex_);
+    auto it = partitionMap_.find(id);
+    if (it != partitionMap_.end()) {
+        if (!storage_->DeletePartition(id)) {
+            return TopoStatusCode::TOPO_STORGE_FAIL;
+        }
+        CopySetKey key(it->second.GetPoolId(), it->second.GetCopySetId());
+        auto ix = copySetMap_.find(key);
+        if (ix != copySetMap_.end()) {
+            ix->second.ReducePartitionNum();
+        }
+        partitionMap_.erase(it);
+        return TopoStatusCode::TOPO_OK;
+    } else {
+        return TopoStatusCode::TOPO_PARTITION_NOT_FOUND;
+    }
+}
+
+TopoStatusCode TopologyImpl::UpdatePartition(const Partition &data) {
+    WriteLockGuard wlockPartition(partitionMutex_);
+    auto it = partitionMap_.find(data.GetPartitionId());
+    if (it != partitionMap_.end()) {
+        if (!storage_->UpdatePartition(data)) {
+            return TopoStatusCode::TOPO_STORGE_FAIL;
+        }
+        it->second = data;
+        return TopoStatusCode::TOPO_OK;
+    } else {
+        return TopoStatusCode::TOPO_ZONE_NOT_FOUND;
+    }
+}
+
+bool TopologyImpl::GetPartition(PartitionIdType partitionId, Partition *out) {
+    ReadLockGuard rlockPartition(partitionMutex_);
+        auto it = partitionMap_.find(partitionId);
+        if (it != partitionMap_.end()) {
+            *out = it->second;
+            return true;
+        }
+        return false;
+}
+
+
+bool TopologyImpl::GetCopysetOfPartition(PartitionIdType id,
+                                         CopySetInfo *out) const {
+    ReadLockGuard rlockPartition(partitionMutex_);
+    auto it = partitionMap_.find(id);
+    if (it != partitionMap_.end()) {
+        PoolIdType poolId = it->second.GetPoolId();
+        CopySetIdType csId = it->second.GetCopySetId();
+        CopySetKey key(poolId, csId);
+        ReadLockGuard rlockCopyset(copySetMutex_);
+        auto iter = copySetMap_.find(key);
+        if (iter != copySetMap_.end()) {
+            *out = iter->second;
+            return true;
+        }
+    }
+    return false;
+}
+
+bool TopologyImpl::GetAvailableCopyset(CopySetInfo *out) const {
+    ReadLockGuard rlockCopySet(copySetMutex_);
+    uint64_t maxPartitionNum = option_.partitionNumberInCopyset;
+    for (auto it : copySetMap_) {
```

Comment:
choose copyset should random

Reply:
> choose copyset should random

fixed: https://github.com/opencurve/curve/blob/17bc2dc556ff18c14058770f771c4138edf189f6/curvefs/src/mds/topology/topology.cpp#L480